albatross/core/traits.h

```diff
@@ -13,268 +13,170 @@
-  static std::true_type test(int);
-  template <typename C> static std::false_type test(...);
+template <typename T, typename FeatureType> class has_possible_fit {
+  template <typename C, typename = decltype(std::declval<C>().fit(
+                            std::declval<std::vector<FeatureType> &>(),
+                            std::declval<MarginalDistribution &>()))>
+  static std::true_type test(C *);
+  template <typename> static std::false_type test(...);
 
 public:
   static constexpr bool value = decltype(test<T>(0))::value;
 };
 
 /*
- * One way to tell the difference between a RegressionModel
- * and a SerializableRegressionModel is by inspecting for a
- * FitType.
- */
-template <typename T> using is_serializable_regression_model = has_fit_type<T>;
-
-/*
- * This traits helper class defines `::type` to be `T::FitType`
- * if a type with that name has been defined for T and will
- * otherwise be `void`.
+ * Determines which object would be returned from a call to:
+ *
+ *   T::get_fit_model(features, targets);
  */
-template <typename T> class fit_type_or_void {
-  template <typename C, typename = typename C::FitType>
-  static typename C::FitType test(int);
-  template <typename C> static void test(...);
+template <typename T, typename FeatureType> class fit_model_type {
+  template <typename C,
+            typename FitModelType = decltype(std::declval<const C>().get_fit_model(
+                std::declval<const std::vector<FeatureType> &>(),
+                std::declval<const MarginalDistribution &>()))>
+  static FitModelType test(C *);
+  template <typename> static void test(...);
 
 public:
+
   typedef decltype(test<T>(0)) type;
 };
 
-/*
- * Helper function for enable_if and is_serializable_regression_model
- */
-template <typename X, typename T>
-using enable_if_serializable =
-    std::enable_if<is_serializable_regression_model<X>::value, T>;
-
-/*
- * Will result in substitution failure if X is not serializable and
- * otherwise resolves to X::FitType.
- */
-template <typename X>
-using fit_type_if_serializable =
-    typename enable_if_serializable<X,
-                                    typename fit_type_or_void<X>::type>::type;
 
 /*
- * The following helper functions let you inspect a type and cereal Archive
- * and determine if the type has a valid serialization method for that Archive
- * type.
+ * fit_type
+ *
+ * Determines which Fit specialization will be returned if you
+ * call ModelType::fit.
  */
-template <typename X, typename Archive> class valid_output_serializer {
-  template <typename T>
-  static typename std::enable_if<
-      1 == cereal::traits::detail::count_output_serializers<T, Archive>::value,
-      std::true_type>::type
-  test(int);
-  template <typename T> static std::false_type test(...);
-
-public:
-  static constexpr bool value = decltype(test<X>(0))::value;
+template <typename ModelType, typename FeatureType, int = 0>
+struct fit_type {
+  typedef void type;
```

Comment:
Yeah, the third specialization of `fit_type` requires it.  It also sort of makes sense for the application since we're asking which fit type would be returned for a model fit with a certain feature type.